### DISCUSSION TITLE
What does scalar run all actually do?

### DISCUSSION THREAD

**Discussion by fazakaslori** [microsoft/git#484] [Q&A] [answered]
Hi guys,

I've been trying to figure out what 'scalar run all' does. In an old blogpost from Derrick (https://devblogs.microsoft.com/devops/introducing-scalar/), and according to some old documentation (https://github.com/microsoft/scalar/blob/master/docs/advanced.md) the command should give some output and run some subroutines. However, when I enter this on a monorepo, it just terminates instantly, without any output.

Does anyone have some info about this?

**Comment by derrickstolee** [comment, marked as answer]
Hi @fazakaslori. You're probably right that we could use more progress output here.

If you are getting a fast termination, then probably something else is going wrong. Could you run it again with `GIT_TRACE2_PERF=1` and share the output?

**Reply by fazakaslori** [reply]
Hmm...It seems we were just unlucky with the timing of the 'scalar run all' command. If not, currently I have no explanation. When I tried to run the command again, 'scalar run all' had all the output, and did what we expected, what the old documentation and blog posts say it does.

To give a little more context, you must know that we have crated some automation processes that generate a lot of git commits on multiple branches. We were doing this to measure how git, and scalar as well, perform on large repos (with a lot of paths, multiple GBs of data and small binary files), with large history (10k+ commits)

In our result with simple git (no scalar), we always saw some commands that took unusually lot of time (because of git gc running sometimes). These unusual results were nowhere to be found on repositories cloned with scalar clone. This lead us to integrating 'scalar run all' into the automation processes, but we still saw no changes in the results, so we tried running it manually but we got no output and the command just terminated. I suspect this was due to running the command close enough in time to when the automation processes terminated. 
Maybe this is about a timing/workload thing, where, if Scalar maintenance has nothing to do, it will just terminate without output?

Do you have some input about this?

**Reply by derrickstolee** [reply]
I do not have an understanding of why you are having trouble here.

Since `scalar run ...` is now just a shim over `git maintenance run --task=<X>`, you could consider skipping Scalar and going straight to the Git commands. That should help identify the actual issue you are having, while also giving you more control over which tasks are run in which order. Scalar runs these tasks:

* `prefetch`
* `incremental-repack`
* `commit-graph`
* `loose-objects`

The only other thing that `scalar run all` does that isn't a `git maintenance` task is `scalar run config` which is the same as `scalar register` to initialize recommended config.
